README.md
# digdag-operator-emr_fleet
[![Jitpack](https://jitpack.io/v/pro.civitaspo/digdag-operator-emr_fleet.svg)](https://jitpack.io/#pro.civitaspo/digdag-operator-emr_fleet) [![CircleCI](https://circleci.com/gh/civitaspo/digdag-operator-emr_fleet.svg?style=shield)](https://circleci.com/gh/civitaspo/digdag-operator-emr_fleet) [![Digdag](https://img.shields.io/badge/digdag-v0.9.27-brightgreen.svg)](https://github.com/treasure-data/digdag/releases/tag/v0.9.27)

This operator is for operating a cluster with instance fleets on Amazon Elastic Map Reduce.

# Overview

- Plugin type: operator

# Usage


```
_export:
  plugin:
    repositories:
      - https://jitpack.io
    dependencies:
      - pro.civitaspo:digdag-operator-emr_fleet:0.0.5
  emr_fleet:
    auth_method: profile

+step1:
  emr_fleet.create_cluster>:
  tags:
    environment: test
  spot_spec:
    block_duration: 1h
  master_fleet:
    bid_percentage: 60
    candidates:
      - instance_type: r3.xlarge
      - instance_type: m3.xlarge
  core_fleet:
    target_capacity: 192
    bid_percentage: 60
    candidates:
      - instance_type: r3.8xlarge
        spot_units: 32
        ebs:
          optimized: false
      - instance_type: r3.4xlarge
        spot_units: 16
  applications: [Hadoop, Spark, Livy]
  configurations:
    - classification: spark-defaults
      properties:
        maximizeResourceAllocation: true
    - classification: capacity-scheduler
      properties:
        yarn.scheduler.capacity.resource-calculator: org.apache.hadoop.yarn.util.resource.DominantResourceCalculator

+step2:
  echo>: ${emr_fleet.last_cluster}

+step3:
  emr_fleet.detect_clusters>:
  created_within: 1h
  regexp: .*${session_uuid}.*

+step4:
  emr_fleet.shutdown_cluster>: ${emr_fleet.last_cluster.id}

+step5:
  emr_fleet.wait_cluster>: ${emr_fleet.last_cluster.id}
  success_states: [TERMINATED]
  _error:
    emr_fleet.shutdown_cluster>: ${emr_fleet.last_cluster.id}
```

# Concept

* Configuable for complicated EMR Cluster
* Available for credentials on servers if server administrators allow the operation
* Small processing that a operator has

# Configuration

## Remarks

- An EMR Cluster is also called an job flow. In the below explanations, I use both terms.
- type `DurationParam` is strings matched `\s*(?:(?<days>\d+)\s*d)?\s*(?:(?<hours>\d+)\s*h)?\s*(?:(?<minutes>\d+)\s*m)?\s*(?:(?<seconds>\d+)\s*s)?\s*`.
  - The strings is used as `java.time.Duration`.

## Common Configuration

### System Options

Define the below options on properties (which is indicated by `-c`, `--config`).

- **emr_fleet.allow_auth_method_env**: Indicates whether users can use **auth_method** `"env"` (boolean, default: `false`)
- **emr_fleet.allow_auth_method_instance**: Indicates whether users can use **auth_method** `"instance"` (boolean, default: `false`)
- **emr_fleet.allow_auth_method_profile**: Indicates whether users can use **auth_method** `"profile"` (boolean, default: `false`)
- **emr_fleet.allow_auth_method_properties**: Indicates whether users can use **auth_method** `"properties"` (boolean, default: `false`)
- **emr_fleet.assume_role_timeout_duration**: Maximum duration which server administer allows when users assume **role_arn**. (`DurationParam`, default: `1h`)

### Secrets

- **emr_fleet.access_key_id**: The AWS Access Key ID to use when submitting EMR jobs. (optional)
- **emr_fleet.secret_access_key**: The AWS Secret Access Key to use when submitting EMR jobs. (optional)
- **emr_fleet.session_token**: The AWS session token to use when submitting EMR jobs. This is used only **auth_method** is `"session"` (optional)
- **emr_fleet.role_arn**: The AWS Role to assume when submitting EMR jobs. (optional)
- **emr_fleet.role_session_name**: The AWS Role Session Name when assuming the role. (default: `digdag-emr_fleet-${session_uuid}`)
- **emr_fleet.http_proxy.host**: proxy host (required if **use_http_proxy** is `true`)
- **emr_fleet.http_proxy.port** proxy port (optional)
- **emr_fleet.http_proxy.scheme** `"https"` or `"http"` (default: `"https"`)
- **emr_fleet.http_proxy.user** proxy user (optional)
- **emr_fleet.http_proxy.password**: http proxy password (optional)

### Options

- **auth_method**: name of mechanism to authenticate requests (`"basic"`, `"env"`, `"instance"`, `"profile"`, `"properties"`, `"anonymous"`, or `"session"`. default: `"basic"`)
  - `"basic"`: uses access_key_id and secret_access_key to authenticate.
  - `"env"`: uses AWS_ACCESS_KEY_ID (or AWS_ACCESS_KEY) and AWS_SECRET_KEY (or AWS_SECRET_ACCESS_KEY) environment variables.
  - `"instance"`: uses EC2 instance profile.
  - `"profile"`: uses credentials written in a file. Format of the file is as following, where `[...]` is a name of profile.
    - **profile_file**: path to a profiles file. (string, default: given by AWS_CREDENTIAL_PROFILES_FILE environment varialbe, or ~/.aws/credentials).
    - **profile_name**: name of a profile. (string, default: `"default"`)
  - `"properties"`: uses aws.accessKeyId and aws.secretKey Java system properties.
  - `"anonymous"`: uses anonymous access. This auth method can access only public files.
  - `"session"`: uses temporary-generated access_key_id, secret_access_key and session_token.
- **use_http_proxy**: Indicate whether using when accessing AWS via http proxy. (boolean, default: `false`)
- **region**: The AWS region to use for EMR service. (string, optional)
- **endpoint**: The AWS EMR endpoint address to use. (string, optional)

## Configuration for `emr_fleet.create_cluster>` operator

### Options

- **name**: The name of the job flow. (string, default: `"digdag-${session_uuid}"`)
- **tags**: A list of tags to associate with a cluster and propagate to Amazon EC2 instances. (string to string map, optional)
- **release_label**: The Amazon EMR release label, which determines the version of open-source application packages installed on the cluster. Release labels are in the form emr-x.x.x, where x.x.x is an Amazon EMR release version. For more information about Amazon EMR release versions and included application versions and features, see [Docs](https://docs.aws.amazon.com/emr/latest/ReleaseGuide/). (string, default: `emr-5.16.0` )
- **custom_ami_id**: Available only in Amazon EMR version 5.7.0 and later. The ID of a custom Amazon EBS-backed Linux AMI. If specified, Amazon EMR uses this AMI when it launches cluster EC2 instances. (string, optional)
- **master_security_groups**: A list of Amazon EC2 security group IDs for the master node. The first one is used as Amazon EMR–managed security group, and the left is used as additional one. See [Docs](https://docs.aws.amazon.com/emr/latest/ManagementGuide/emr-security-groups.html) to get more information. (array of string, optional)
- **slave_security_groups**: A list of Amazon EC2 security group IDs for the slave node. The first one is used as Amazon EMR–managed security group, and the left is used as additional one. See [Docs](https://docs.aws.amazon.com/emr/latest/ManagementGuide/emr-security-groups.html) to get more information. (array of string, optional)
- **ssh_key**: The name of the EC2 key pair that can be used to ssh to the nodes. (string, optional)
- **subnet_ids**: Applies to clusters that use the instance fleet configuration. When multiple EC2 subnet IDs are specified, Amazon EMR evaluates them and launches instances in the optimal subnet. (array of string, optional)
- **availability_zones**: When multiple Availability Zones are specified, Amazon EMR evaluates them and launches instances in the optimal Availability Zone. (array of string, optional)
- **spot_spec**: The launch specification for Spot instances in the instance fleet, which determines the defined duration and provisioning timeout behavior. (map, default: `{}`)
  - **block_duration**: The defined duration for Spot instances (also known as Spot blocks). When specified, the Spot instance does not terminate before the defined duration expires, and defined duration pricing for Spot instances applies. Current valid values are `"1h"`, `"2h"`, `"3h"`, `"4h"`, `"5h"`, or `"6h"`. The duration period starts as soon as a Spot instance receives its instance ID. At the end of the duration, Amazon EC2 marks the Spot instance for termination and provides a Spot instance termination notice, which gives the instance a two-minute warning before it terminates. (`DurationParam`, optional)
  - **timeout_action**: The action to take when **target_spot_capacity** has not been fulfilled when the **timeout_duration** has expired. Spot instances are not uprovisioned within the Spot provisioining timeout. Valid values are `TERMINATE_CLUSTER` and `SWITCH_TO_ON_DEMAND`. `SWITCH_TO_ON_DEMAND` specifies that if no Spot instances are available, On-Demand Instances should be provisioned to fulfill any remaining Spot capacity. (string, default: `TERMINATE_CLUSTER`)
  - **timeout_duration**: The spot provisioning timeout period. If Spot instances are not provisioned within this time period, the **timeout_action** is taken. Minimum value is `"5m"` and maximum value is `"1d"`. The timeout applies only during initial provisioning, when the cluster is first created. (`DurationParam`, default: `45m`)
- **master_fleet**: Describes the EC2 instances and instance configurations for master node that use the instance fleet configuration. (map, required)
  - **name**: The friendly name of the instance fleet. (string, default: `master instance fleet`)
  - **use_spot_instance**: Indicates whether master node uses Spot instance or On-Demand instance. (boolean, default: `true`)
  - **bid_percentage**: The bid price, as a percentage of On-Demand price, for each EC2 Spot instance as defined by **instance_type**. Expressed as a number (for example, 20 specifies 20%). This value is used for **candidates**'s **bid_percentage** default value. (double, default: `100.0`)
  - **candidates**: An instance type configuration for each instance type in an instance fleet, which determines the EC2 instances Amazon EMR attempts to provision to fulfill On-Demand and Spot target capacities. There can be a maximum of 5 instance type configurations in a fleet. (array of map, required)
    - **bid_price**: The bid price for each EC2 Spot instance type as defined by **instance_type**. Expressed in USD. If neither **bid_price** nor **bid_percentage** is provided, **bid_percentage** defaults to 100%. (string, optional)
    - **bid_percentage**: The bid price, as a percentage of On-Demand price, for each EC2 Spot instance as defined by **instance_type**. Expressed as a number (for example, 20 specifies 20%). If neither **bid_price** nor **bid_percentage** is provided, **bid_percentage** defaults to 100%. (double, default: `${master_fleet.bid_percentage}`)
    - **instance_type**: An EC2 instance type, such as `m3.xlarge`. (string, required)
    - **configurations**: A configuration classification that applies when provisioning cluster instances, which can include configurations for applications and software that run on the cluster. (array of map, optional)
      - **classification**: The classification within a configuration. (string, required)
      - **properties**: A set of properties specified within a configuration classification. (string to string map, default: optional)
      - **configurations**: A list of additional configurations to apply within a configuration object. (array of this configurations object, optional)
    - **ebs**: The Amazon EBS configuration of a cluster instance. (map, default: `{}`)
      - **optimized**: Indicates whether an Amazon EBS volume is EBS-optimized. (boolean, default: `true`)
      - **iops**: The number of I/O operations per second (IOPS) that the volume supports. (integer, optional)
      - **size**: The volume size, in gibibytes (GiB). This can be a number from 1 - 1024. If the volume type is EBS-optimized, the minimum value is 10. (integer, default: `256`)
      - **type**: The volume type. Volume types supported are gp2, io1, standard. (string, default: `gp2`)
      - **volumes_per_instance**: Number of EBS volumes with a specific volume configuration that will be associated with every instance in the instance group. (integer, default: `1`)
- **core_fleet**: Describes the EC2 instances and instance configurations for core nodes that use the instance fleet configuration. (map, required)
  - **name**: The friendly name of the instance fleet. (string, default: `core instance fleet`)
  - **target_capacity**: The target capacity of **spot_units** for the instance fleet, which determines how many Spot instances to provision. When the instance fleet launches, Amazon EMR tries to provision Spot instances as specified by **candidates** settings. Each instance configuration has a specified **spot_units**. When a Spot instance is provisioned, the **spot_units** count toward the target capacity. Amazon EMR provisions instances until the target capacity is totally fulfilled, even if this results in an overage. (integer, required)
  - **bid_percentage**: The bid price, as a percentage of On-Demand price, for each EC2 Spot instance as defined by **instance_type**. Expressed as a number (for example, 20 specifies 20%). This value is used for **candidates**'s **bid_percentage** default value. (double, default: `100.0`)
  - **candidates**: An instance type configuration for each instance type in an instance fleet, which determines the EC2 instances Amazon EMR attempts to provision to fulfill On-Demand and Spot target capacities. There can be a maximum of 5 instance type configurations in a fleet. (array of map, required)
    - **bid_price**: The bid price for each EC2 Spot instance type as defined by **instance_type**. Expressed in USD. If neither **bid_price** nor **bid_percentage** is provided, **bid_percentage** defaults to 100%. (string, optional)
    - **bid_percentage**: The bid price, as a percentage of On-Demand price, for each EC2 Spot instance as defined by **instance_type**. Expressed as a number (for example, 20 specifies 20%). If neither **bid_price** nor **bid_percentage** is provided, **bid_percentage** defaults to 100%. (double, default: `${core_fleet.bid_percentage}`)
    - **instance_type**: An EC2 instance type, such as `m3.xlarge`. (string, required)
    - **configurations**: A configuration classification that applies when provisioning cluster instances, which can include configurations for applications and software that run on the cluster. (array of map, optional)
      - **classification**: The classification within a configuration. (string, required)
      - **properties**: A set of properties specified within a configuration classification. (string to string map, optional)
      - **configurations**: A list of additional configurations to apply within a configuration object. (array of this configurations object, optional)
    - **ebs**: The Amazon EBS configuration of a cluster instance. (map, default: `{}`)
      - **optimized**: Indicates whether an Amazon EBS volume is EBS-optimized. (boolean, default: `true`)
      - **iops**: The number of I/O operations per second (IOPS) that the volume supports. (integer, optional)
      - **size**: The volume size, in gibibytes (GiB). This can be a number from 1 - 1024. If the volume type is EBS-optimized, the minimum value is 10. (integer, default: `256`)
      - **type**: The volume type. Volume types supported are gp2, io1, standard. (string, default: `gp2`)
      - **volumes_per_instance**: Number of EBS volumes with a specific volume configuration that will be associated with every instance in the instance group. (integer, default: `1`)
    - **spot_units**: The number of units that a provisioned instance of this type provides toward fulfilling the **target_capacity**. This value must be 1 or greater. (integer, default: `1`)
- **task_fleet**: Describes the EC2 instances and instance configurations for task nodes that use the instance fleet configuration. (map, optional)
  - **name**: The friendly name of the instance fleet. (string, default: `task instance fleet`)
  - The other options are the same as **task** option.
- **log_uri**: The location in Amazon S3 to write the log files of the job flow. If a value is not provided, logs are not created. (string, optional)
- **additional_info**: A JSON string for selecting additional features. (string, optional)
- **visible**: Whether the cluster is visible to all IAM users of the AWS account associated with the cluster. If this value is set to true, all IAM users of that AWS account can view and (if they have the proper policy permissions set) manage the cluster. If it is set to false, only the IAM user that created the cluster can view and manage it. (boolean, default: `true`)
- **security_configuration**: The name of a security configuration to apply to the cluster. (string, optional)
- **instance_profile**: Also called job flow role and EC2 role. An IAM role for an EMR cluster. The EC2 instances of the cluster assume this role. The default role is EMR_EC2_DefaultRole. In order to use the default role, you must have already created it using the CLI or console. (string, default: `EMR_EC2_DefaultRole`)
- **service_role**: The IAM role that will be assumed by the Amazon EMR service to access AWS resources on your behalf. (string, default: `EMR_DefaultRole`)

CHANGELOG.md
0.0.5 (2018-09-09)
==================

* [New Feature] Add `content` option for bootstrap action that uploads content to `script` location.
* [New Feature] Add `run_if` option for bootstrap action that wraps `run-if` script provided by EMR for easier use.

0.0.4 (2018-08-09)
==================

* [Enhancement] Fail correctly if not retryable.

0.0.3 (2018-08-06)
==================

* [Enhancement] Enable to merge emr_fleet.${task_name} exports.

0.0.2 (2018-08-06)
==================

* [Destructive Change] Change `bootstrap_action` configuration specification.
* [Enhancement] Use `RetryExecutor` of digdag-plugin-utils instead of original one.


0.0.1 (2018-07-31)
==================

* First Release


src/main/scala/pro/civitaspo/digdag/plugin/emr_fleet/operator/EmrFleetCreateClusterOperator.scala
package pro.civitaspo.digdag.plugin.emr_fleet.operator

import java.net.URI
import java.nio.charset.StandardCharsets.UTF_8

import com.amazonaws.services.elasticmapreduce.model.{
  Application,
  BootstrapActionConfig,
  ClusterState,
  Configuration,
  EbsBlockDeviceConfig,
  EbsConfiguration,
  InstanceFleetConfig,
  InstanceFleetProvisioningSpecifications,
  InstanceFleetType,
  InstanceTypeConfig,
  JobFlowInstancesConfig,
  PlacementType,
  RunJobFlowRequest,
  ScriptBootstrapActionConfig,
  SpotProvisioningSpecification,
  SpotProvisioningTimeoutAction,
  Tag,
  VolumeSpecification
}
import com.amazonaws.services.elasticmapreduce.model.ClusterState.{RUNNING, TERMINATED, TERMINATED_WITH_ERRORS, TERMINATING, WAITING}
import com.amazonaws.services.elasticmapreduce.model.InstanceFleetType.{CORE, MASTER, TASK}
import com.amazonaws.services.elasticmapreduce.model.SpotProvisioningTimeoutAction.TERMINATE_CLUSTER
import com.amazonaws.services.s3.AmazonS3URI
import com.google.common.base.Optional
import com.google.common.collect.ImmutableList
import io.digdag.client.config.{Config, ConfigException, ConfigKey}
import io.digdag.spi.{OperatorContext, TaskResult, TemplateEngine}
import io.digdag.util.DurationParam

import scala.collection.JavaConverters._
import scala.util.{Random, Try}
import scala.util.hashing.MurmurHash3

class EmrFleetCreateClusterOperator(operatorName: String, context: OperatorContext, systemConfig: Config, templateEngine: TemplateEngine)
    extends AbstractEmrFleetOperator(operatorName, context, systemConfig, templateEngine) {

  object AmazonS3URI {
    def apply(path: String): AmazonS3URI = new AmazonS3URI(path, false)
  }

  // ref. https://docs.aws.amazon.com/emr/latest/ManagementGuide/emr-plan-bootstrap.html#emr-bootstrap-runif
  protected val RUN_IF_SCRIPT_LOCATION: String = "s3://elasticmapreduce/bootstrap-actions/run-if"

  protected val clusterName: String = params.get("name", classOf[String], s"digdag-${params.get("session_uuid", classOf[String])}")
  protected val tags: Map[String, String] = params.getMapOrEmpty("tags", classOf[String], classOf[String]).asScala.toMap
  protected val releaseLabel: String = params.get("release_label", classOf[String], "emr-5.16.0")
  protected val customAmiId: Optional[String] = params.getOptional("custom_ami_id", classOf[String])
  protected val masterSecurityGroups: Seq[String] = params.getListOrEmpty("master_security_groups", classOf[String]).asScala
  protected val slaveSecurityGroups: Seq[String] = params.getListOrEmpty("slave_security_groups", classOf[String]).asScala
  protected val sshKey: Optional[String] = params.getOptional("ssh_key", classOf[String])
  protected val subnetIds: Seq[String] = params.getListOrEmpty("subnet_ids", classOf[String]).asScala
  protected val availabilityZones: Seq[String] = params.getListOrEmpty("availability_zones", classOf[String]).asScala
  protected val spotSpec: Config = params.getNestedOrGetEmpty("spot_spec")
  protected val masterFleet: Config = params.getNested("master_fleet")
  protected val coreFleet: Config = params.getNested("core_fleet")
  protected val taskFleet: Config = params.getNestedOrGetEmpty("task_fleet")
  protected val logUri: Optional[String] = params.getOptional("log_uri", classOf[String])
  protected val additionalInfo: Optional[String] = params.getOptional("additional_info", classOf[String])
  protected val isVisible: Boolean = params.get("visible", classOf[Boolean], true)
  protected val securityConfiguration: Optional[String] = params.getOptional("security_configuration", classOf[String])
  protected val instanceProfile: String = params.get("instance_profile", classOf[String], "EMR_EC2_DefaultRole")
  protected val serviceRole: String = params.get("service_role", classOf[String], "EMR_DefaultRole")
  protected val applications: Seq[String] = params.getListOrEmpty("applications", classOf[String]).asScala
  protected val applicationConfigurations: Seq[Config] = params.getListOrEmpty("configurations", classOf[Config]).asScala
  protected val bootstrapActions: Seq[Config] = params.getListOrEmpty("bootstrap_actions", classOf[Config]).asScala
  protected val keepAliveWhenNoSteps: Boolean = params.get("keep_alive_when_no_steps", classOf[Boolean], true)
  protected val terminationProtected: Boolean = params.get("termination_protected", classOf[Boolean], false)
  protected val waitAvailableState: Boolean = params.get("wait_available_state", classOf[Boolean], true)
  protected val waitTimeoutDuration: DurationParam = params.get("wait_timeout_duration", classOf[DurationParam], DurationParam.parse("45m"))

  protected lazy val configureInstanceFleetProvisioningSpecifications: InstanceFleetProvisioningSpecifications = {
    val blockDuration: Optional[DurationParam] = spotSpec.getOptional("block_duration", classOf[DurationParam])
    val timeoutAction: SpotProvisioningTimeoutAction = spotSpec.get("timeout_action", classOf[SpotProvisioningTimeoutAction], TERMINATE_CLUSTER)
    val timeoutDuration: DurationParam = spotSpec.get("timeout_duration", classOf[DurationParam], DurationParam.parse("45m"))

    val s = new SpotProvisioningSpecification()
    if (blockDuration.isPresent) {
      val bd: Int = blockDuration.get().getDuration.toMinutes.toInt
      if (!Seq[Int](1, 2, 3, 4, 5, 6).map(_ * 60).contains(bd)) {
        logger.warn(
          s"""[$operatorName] "1h", "2h", "3h", "4h", "5h", or "6h" are allowed for `block_duration`, so "${blockDuration.get().toString}" is invalid."""
        )
        logger.warn(
          s"""[$operatorName] `$operatorName` operator respects the options you set, but the behaviour depends on AWS. See the document ( https://docs.aws.amazon.com/emr/latest/ManagementGuide/emr-instance-fleet.html )"""
        )
      }
      s.setBlockDurationMinutes(blockDuration.get().getDuration.toMinutes.toInt)
    }
    s.setTimeoutAction(timeoutAction)
    s.setTimeoutDurationMinutes(timeoutDuration.getDuration.toMinutes.toInt)

    new InstanceFleetProvisioningSpecifications().withSpotSpecification(s)
  }

  protected def configureMasterFleet: InstanceFleetConfig = {
    val name: String = masterFleet.get("name", classOf[String], "master instance fleet")
    val useSpotInstance: Boolean = masterFleet.get("use_spot_instance", classOf[Boolean], true)
    val defaultBidPercentage: Double = masterFleet.get("bid_percentage", classOf[Double], 100.0)
    val candidates: Seq[Config] = masterFleet.getList("candidates", classOf[Config]).asScala

    val c = new InstanceFleetConfig()
    c.setInstanceFleetType(MASTER)
    c.setName(name)
    c.setLaunchSpecifications(configureInstanceFleetProvisioningSpecifications)
    if (useSpotInstance) {
      c.setTargetSpotCapacity(1)
      c.setTargetOnDemandCapacity(0)
    }
    else {
      c.setTargetSpotCapacity(0)
      c.setTargetOnDemandCapacity(1)
    }
    c.setInstanceTypeConfigs(seqAsJavaList(candidates.map(configureCandidate(_, defaultBidPercentage))))
    c
  }

  protected def configureSlaveFleet(fleetType: InstanceFleetType, fleetConfiguration: Config): InstanceFleetConfig = {
    val name: String = fleetConfiguration.get("name", classOf[String], s"${fleetType.toString.toLowerCase} instance fleet")
    val targetCapacity: Int = fleetConfiguration.get("target_capacity", classOf[Int])
    val defaultBidPercentage: Double = masterFleet.get("bid_percentage", classOf[Double], 100.0)
    val candidates: Seq[Config] = fleetConfiguration.getList("candidates", classOf[Config]).asScala

    new InstanceFleetConfig()
      .withInstanceFleetType(fleetType)
      .withName(name)
      .withLaunchSpecifications(configureInstanceFleetProvisioningSpecifications)
      .withTargetSpotCapacity(targetCapacity)
      .withInstanceTypeConfigs(seqAsJavaList(candidates.map(configureCandidate(_, defaultBidPercentage))))
  }

  protected def configureCandidate(candidate: Config, defaultBidPercentage: Double): InstanceTypeConfig = {
    val bidPrice: Optional[String] = candidate.getOptional("bid_price", classOf[String])
    val bidPercentage: Double = candidate.get("bid_percentage", classOf[Double], defaultBidPercentage)
    val instanceType: String = candidate.get("instance_type", classOf[String])
    val applicationConfigurations: Seq[Config] = candidate.getListOrEmpty("configurations", classOf[Config]).asScala
    val ebs: Config = candidate.getNestedOrGetEmpty("ebs")
    val spotUnits: Int = candidate.get("spot_units", classOf[Int], 1)

    val c = new InstanceTypeConfig()
    if (bidPrice.isPresent) c.setBidPrice(bidPrice.get())
    c.setBidPriceAsPercentageOfOnDemandPrice(bidPercentage)
    c.setInstanceType(instanceType)
    c.setConfigurations(seqAsJavaList(applicationConfigurations.map(configureApplicationConfiguration)))
    c.setEbsConfiguration(configureEbs(ebs))
    c.setWeightedCapacity(spotUnits)
    c
  }

  protected def configureEbs(ebs: Config): EbsConfiguration = {
    val isOptimized: Boolean = ebs.get("optimized", classOf[Boolean], true)
    val iops: Optional[Int] = ebs.getOptional("iops", classOf[Int])
    val size: Int = ebs.get("size", classOf[Int], 256)
    val volumeType: String = ebs.get("type", classOf[String], "gp2")
    val volumesPerInstance: Int = ebs.get("volumes_per_instance", classOf[Int], 1)

    val volumeSpecification = new VolumeSpecification()
    if (iops.isPresent) volumeSpecification.setIops(iops.get())
    volumeSpecification.setSizeInGB(size)
    volumeSpecification.setVolumeType(volumeType)

    new EbsConfiguration()
      .withEbsOptimized(isOptimized)
      .withEbsBlockDeviceConfigs(
        new EbsBlockDeviceConfig()
          .withVolumesPerInstance(volumesPerInstance)
          .withVolumeSpecification(volumeSpecification)
      )
  }

  protected def configureApplicationConfiguration(applicationConfiguration: Config): Configuration = {
    val ac = applicationConfiguration // to shorten var name
    val classification: String = ac.get("classification", classOf[String])
    val properties: Map[String, String] = ac.getMapOrEmpty("properties", classOf[String], classOf[String]).asScala.toMap
    val configurations: Seq[Config] = ac.getListOrEmpty("configurations", classOf[Config]).asScala

    val c = new Configuration()
    c.setClassification(classification)
    if (properties.nonEmpty) c.setProperties(mapAsJavaMap(properties))
    if (configurations.nonEmpty) c.setConfigurations(seqAsJavaList(configurations.map(configureApplicationConfiguration)))
    c
  }

  protected def configureBootstrapAction(bootstrapAction: Config): Seq[BootstrapActionConfig] = {
    val name: String = bootstrapAction.get("name", classOf[String])
    val script: String = bootstrapAction.get("script", classOf[String])
    val args: Seq[String] = bootstrapAction.getListOrEmpty("args", classOf[String]).asScala
    val contentOrFile: Optional[String] = bootstrapAction.getOptional("content", classOf[String])
    val runIf: Optional[String] = bootstrapAction.getOptional("run_if", classOf[String])

    if (contentOrFile.isPresent) uploadBootstrapActionScript(contentOrFile.get(), script)

    val builder = Seq.newBuilder[BootstrapActionConfig]
    if (runIf.isPresent) builder ++= buildRunIfBootstrapAction(runIf.get(), name, script, args)
    else builder += new BootstrapActionConfig(name, new ScriptBootstrapActionConfig(script, seqAsJavaList(args)))

    builder.result()
  }

  protected def uploadBootstrapActionScript(contentOrFile: String, script: String): Unit = {
    if (!script.startsWith("s3://")) throw new ConfigException(s"[$operatorName] `script` must start with 's3://' if uploading content as bootstrap action.")
    val s3Uri = AmazonS3URI(script)
    val content = Try(workspace.templateFile(templateEngine, workspace.getFile(contentOrFile).getPath, UTF_8, params)).getOrElse(contentOrFile)
    logger.info(s"[$operatorName] Upload content to $script. (content hash (MurmurHash3): ${MurmurHash3.bytesHash(content.getBytes, 1234)})")
    withS3(_.putObject(s3Uri.getBucket, s3Uri.getKey, content))
  }

  protected def buildRunIfBootstrapAction(condition: String, name: String, script: String, args: Seq[String]): Seq[BootstrapActionConfig] = {
    val builder = Seq.newBuilder[BootstrapActionConfig]
    new URI(script) match {
      case uri if uri.getScheme.contentEquals("file") =>
        builder += new BootstrapActionConfig()
          .withName(s"$name if $condition")
          .withScriptBootstrapAction(
            new ScriptBootstrapActionConfig()
              .withPath(RUN_IF_SCRIPT_LOCATION)
              .withArgs(Seq(condition, uri.getPath) ++ args: _*)
          )
      case uri if uri.getScheme.contentEquals("s3") =>
        // NOTE: `run-if` script provided by EMR has a bug: https://github.com/aws-samples/emr-bootstrap-actions/pull/211
        //       so, download script first, then execute it.
        val localFileName: String = s"digdag-${params.get("session_uuid", classOf[String])}-${Random.alphanumeric.take(10).mkString}"
        builder += new BootstrapActionConfig()
          .withName(s"$name if $condition: Download $script as /tmp/$localFileName")
          .withScriptBootstrapAction(
            new ScriptBootstrapActionConfig()
              .withPath(RUN_IF_SCRIPT_LOCATION)
              .withArgs(Seq(condition, "aws", "s3", "cp", script, s"/tmp/$localFileName"): _*)
          )
        builder += new BootstrapActionConfig()
          .withName(s"$name if $condition: chmod 755 /tmp/$localFileName ($script)")
          .withScriptBootstrapAction(
            new ScriptBootstrapActionConfig()
              .withPath(RUN_IF_SCRIPT_LOCATION)
              .withArgs(Seq(condition, "chmod", "755", s"/tmp/$localFileName"): _*)
          )
        builder += new BootstrapActionConfig()
          .withName(s"$name if $condition: Execute /tmp/$localFileName ($script)")
          .withScriptBootstrapAction(
            new ScriptBootstrapActionConfig()
              .withPath(RUN_IF_SCRIPT_LOCATION)
              .withArgs(Seq(condition, s"/tmp/$localFileName") ++ args: _*)
          )
      case _ =>
        throw new ConfigException(s"[$operatorName] Invalid bootstrap action path, must be a location in Amazon S3 or a local path starting with 'file:'.")
    }
    builder.result()
  }

  protected def configureInstances: JobFlowInstancesConfig = {
    val c = new JobFlowInstancesConfig()

    if (masterSecurityGroups.nonEmpty) {
      c.setEmrManagedMasterSecurityGroup(masterSecurityGroups.head)
      if (masterSecurityGroups.tail.nonEmpty) {
        c.setAdditionalMasterSecurityGroups(seqAsJavaList(masterSecurityGroups.tail))
      }
    }
    if (slaveSecurityGroups.nonEmpty) {
      c.setEmrManagedSlaveSecurityGroup(slaveSecurityGroups.head)
      if (slaveSecurityGroups.tail.nonEmpty) {
        c.setAdditionalSlaveSecurityGroups(seqAsJavaList(slaveSecurityGroups.tail))
      }
    }
    if (availabilityZones.nonEmpty) c.setPlacement(new PlacementType().withAvailabilityZones(availabilityZones: _*))
    if (sshKey.isPresent) c.setEc2KeyName(sshKey.get())
    if (subnetIds.nonEmpty) c.setEc2SubnetIds(seqAsJavaList(subnetIds))

    val instanceTypeConfigsBuilder = Seq.newBuilder[InstanceFleetConfig]
    instanceTypeConfigsBuilder += configureMasterFleet
    instanceTypeConfigsBuilder += configureSlaveFleet(CORE, coreFleet)
    if (!taskFleet.isEmpty) instanceTypeConfigsBuilder += configureSlaveFleet(TASK, taskFleet)
    c.setInstanceFleets(seqAsJavaList(instanceTypeConfigsBuilder.result()))

    c.setKeepJobFlowAliveWhenNoSteps(keepAliveWhenNoSteps)
    c.setTerminationProtected(terminationProtected)

    c
  }

  protected def buildCreateClusterRequest: RunJobFlowRequest = {
    new RunJobFlowRequest()
      .withAdditionalInfo(additionalInfo.orNull)
      .withApplications(applications.map(a => new Application().withName(a)): _*)
      .withBootstrapActions(bootstrapActions.flatMap(configureBootstrapAction): _*)
      .withConfigurations(applicationConfigurations.map(configureApplicationConfiguration): _*)
      .withCustomAmiId(customAmiId.orNull)
      .withJobFlowRole(instanceProfile)
      .withLogUri(logUri.orNull)
      .withName(clusterName)
      .withReleaseLabel(releaseLabel)
      .withSecurityConfiguration(securityConfiguration.orNull)
      .withServiceRole(serviceRole)
      .withTags(tags.toSeq.map(m => new Tag().withKey(m._1).withValue(m._2)): _*)
      .withVisibleToAllUsers(isVisible)
      .withInstances(configureInstances)
  }

  override def runTask(): TaskResult = {
    val req = buildCreateClusterRequest
    val r = withEmr { emr =>
      emr.runJobFlow(req)
    }
    logger.info(s"""[$operatorName] The request to create a cluster is accepted: ${r.getJobFlowId}, request: $req""")

    val p = newEmptyParams
    p.getNestedOrSetEmpty("emr_fleet").getNestedOrSetEmpty("last_cluster").set("id", r.getJobFlowId)

    val builder = TaskResult.defaultBuilder(request)
    builder.storeParams(p)
    builder.resetStoreParams(ImmutableList.of(ConfigKey.of("emr_fleet", "last_cluster")))
    if (waitAvailableState) {
      logger.info(s"""[$operatorName] run a sub task: emr_fleet.wait_cluster""")
      builder.subtaskConfig(buildWaiterSubTaskConfig(r.getJobFlowId))
    }
    builder.build()
  }

  protected def buildWaiterSubTaskConfig(clusterId: String): Config = {
    val p = newEmptyParams
    p.set("_command", clusterId)
    p.set("_type", "emr_fleet.wait_cluster")
    p.set("success_states", seqAsJavaList(Seq[ClusterState](RUNNING, WAITING)))
    p.set("error_states", seqAsJavaList(Seq[ClusterState](TERMINATED, TERMINATED_WITH_ERRORS, TERMINATING)))
    p.set("timeout_duration", waitTimeoutDuration.toString)

    p.set("auth_method", authMethod)
    p.set("profile_name", profileName)
    if (profileFile.isPresent) p.set("profile_file", profileFile.get())
    p.set("use_http_proxy", useHttpProxy)
    if (region.isPresent) p.set("region", region.get())
    if (endpoint.isPresent) p.set("endpoint", endpoint.get())

    p
  }
}
